homeassistant/components/insteon/fan.py

```diff
@@ -1,56 +1,49 @@
 """Support for INSTEON fans via PowerLinc Modem."""
 import logging
 
+from pyinsteon.constants import FanSpeed
+
 from homeassistant.components.fan import (
+    DOMAIN,
     SPEED_HIGH,
     SPEED_LOW,
     SPEED_MEDIUM,
     SPEED_OFF,
     SUPPORT_SET_SPEED,
     FanEntity,
 )
-from homeassistant.const import STATE_OFF
 
 from .insteon_entity import InsteonEntity
+from .utils import async_add_insteon_entities
 
 _LOGGER = logging.getLogger(__name__)
 
-SPEED_TO_HEX = {SPEED_OFF: 0x00, SPEED_LOW: 0x3F, SPEED_MEDIUM: 0xBE, SPEED_HIGH: 0xFF}
-
-FAN_SPEEDS = [STATE_OFF, SPEED_LOW, SPEED_MEDIUM, SPEED_HIGH]
-
 
 async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up the INSTEON device class for the hass platform."""
-    insteon_modem = hass.data["insteon"].get("modem")
-
-    address = discovery_info["address"]
-    device = insteon_modem.devices[address]
-    state_key = discovery_info["state_key"]
-
-    _LOGGER.debug(
-        "Adding device %s entity %s to Fan platform",
-        device.address.hex,
-        device.states[state_key].name,
+    async_add_insteon_entities(
+        hass, DOMAIN, InsteonFan, async_add_entities, discovery_info
     )
 
-    new_entity = InsteonFan(device, state_key)
-
-    async_add_entities([new_entity])
-
 
 class InsteonFan(InsteonEntity, FanEntity):
     """An INSTEON fan component."""
 
     @property
     def speed(self) -> str:
         """Return the current speed."""
-        return self._hex_to_speed(self._insteon_device_state.value)
+        if self._insteon_device_group.value == FanSpeed.HIGH:
+            return SPEED_HIGH
+        if self._insteon_device_group.value == FanSpeed.MEDIUM:
+            return SPEED_MEDIUM
+        if self._insteon_device_group.value == FanSpeed.LOW:
+            return SPEED_LOW
+        return SPEED_OFF
 
     @property
     def speed_list(self) -> list:
         """Get the list of available speeds."""
-        return FAN_SPEEDS
+        return [str(speed) for speed in FanSpeed]
```

Comment:
The speed list should return the available home assistant speed constants that we return from the `speed` property.